sdk/servicebus/azure-servicebus/azure/servicebus/_common/_connection_string_parser.py

```diff
@@ -0,0 +1,67 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse  # type: ignore
+
+from azure.servicebus.management._models import DictMixin
+
+class ServiceBusConnectionStringProperties(DictMixin):
+    """
+    Properties of a connection string.
+    """
+    def __init__(self, **kwargs):
+        self.fully_qualified_namespace = kwargs.pop('fully_qualified_namespace', None)
+        self.endpoint = kwargs.pop('endpoint', None)
+        self.entity_path = kwargs.pop('entity_path', None)
+        self.shared_access_signature = kwargs.pop('shared_access_signature', None)
+        self.shared_access_key_name = kwargs.pop('shared_access_key_name', None)
+        self.shared_access_key = kwargs.pop('shared_access_key', None)
+
+
+class ServiceBusConnectionStringParser(object):
+    """Parse the connection string.
+
+    :param conn_str: The connection string that has to be parsed.
+    """
+    def __init__(self, conn_str):
+        # type: (str) -> None
+        """
+        :param conn_str: The connection string to parse.
+        :type conn_str: str
+        """
+        self._conn_str = conn_str
+
+    def parse(self):
+        # type() -> ServiceBusConnectionStringProperties
+        """
+        Parse the connection string.
+        """
+        conn_str = self._conn_str.rstrip(";")
+        conn_settings = [s.split("=", 1) for s in conn_str.split(";")]
+        if any(len(tup) != 2 for tup in conn_settings):
+            raise ValueError("Connection string is either blank or malformed.")
+        conn_settings = dict(conn_settings)
+        shared_access_signature = conn_settings.get('SharedAccessSignature')
+        shared_access_key = conn_settings.get('SharedAccessKey')
+        if shared_access_signature is not None and shared_access_key is not None:
+            raise ValueError("Only one of the SharedAccessKey or SharedAccessSignature must be present.")
```

Comment:
should we also raise error if only one of `SharedAccessKey` or `SharedAccessKeyName` is provided (another is missing)?

Reply:
I'm actually wondering whether we want the validation here or not.

We're already talking about artificial connection string flows, what's to say they aren't further truncated?  Should we allow that?

@rakshith91  was there any mention of this in the initial spec that Jesse wrote up?  If not, this may be worth unifying on, since to @yunhaoling 's point that check _does_ technically exist elsewhere (e.g. in the Credential object flow).

In fewer words, I don't know if we decided on "Fail fast" or "make this more robust" and if it's not in the spec we should nail that down.

Reply:
Yes, the issue that Jesse logged does talk about validation. See #14574

In JS, my initial approach was to not have any validation and simply return a dictionary of all the key/value pairs in the connection string. You can see the discussion for this in https://github.com/Azure/azure-sdk-for-js/pull/11909

But, what we landed on was a helper method that takes the connection string, performs validation and returns a strongly typed output. See https://github.com/Azure/azure-sdk-for-js/pull/11909

In terms of validation, you can see what we did in JS here: https://github.com/Azure/azure-sdk-for-js/blob/112bb1da8c66ba90655413e3a76c0a899f2d1dee/sdk/servicebus/service-bus/src/util/connectionStringUtils.ts#L63-L81

Reply:
Additionally, here's a good write up by Jesse
https://gist.github.com/jsquire/13742f53bf642ec67baf9491fc844cff